tests/hierarchy_driver.cc

```diff
@@ -63,33 +69,49 @@ void matrix_free_two_grids(std::shared_ptr<boost::property_tree::ptree> params)
       std::make_shared<TestMFMeshEvaluator<dim, fe_degree, double>>(
           mf_laplace._dof_handler, mf_laplace._constraints,
           mf_laplace._laplace_operator, material_property);
-  mfmg::Hierarchy<DVector> hierarchy(comm, evaluator, params, timer);
-
-  // We want to do 20 V-cycle iterations. The rhs of is zero.
-  // Use D(istributed)Vector because deal has its own Vector class
-  DVector residual(rhs);
-  unsigned int const n_cycles = 20;
-  std::vector<double> res(n_cycles + 1);
 
-  mf_laplace._laplace_operator.vmult(residual, solution);
-  residual.sadd(-1., 1., rhs);
-  auto const residual0_norm = residual.l2_norm();
+  mfmg::Hierarchy<DVector> hierarchy(comm, evaluator, params, timer);
 
-  std::cout << std::scientific;
-  res[0] = 1.0;
-  for (unsigned int i = 0; i < n_cycles; ++i)
+  if (!test_preconditioner)
   {
-    hierarchy.vmult(solution, rhs);
+    // We want to do 20 V-cycle iterations. The rhs of is zero.
+    // Use D(istributed)Vector because deal has its own Vector class
+    DVector residual(rhs);
+    unsigned int const n_cycles = 20;
+    std::vector<double> res(n_cycles + 1);
 
     mf_laplace._laplace_operator.vmult(residual, solution);
     residual.sadd(-1., 1., rhs);
-    double rel_residual = residual.l2_norm() / residual0_norm;
-    res[i + 1] = rel_residual;
-  }
+    auto const residual0_norm = residual.l2_norm();
 
-  double const conv_rate = res[n_cycles] / res[n_cycles - 1];
-  pcout << "Convergence rate: " << std::fixed << std::setprecision(2)
-        << conv_rate << std::endl;
+    std::cout << std::scientific;
+    res[0] = 1.0;
+    for (unsigned int i = 0; i < n_cycles; ++i)
+    {
+      hierarchy.vmult(solution, rhs);
+
+      mf_laplace._laplace_operator.vmult(residual, solution);
+      residual.sadd(-1., 1., rhs);
+      double rel_residual = residual.l2_norm() / residual0_norm;
+      res[i + 1] = rel_residual;
+    }
+
+    double const conv_rate = res[n_cycles] / res[n_cycles - 1];
+    pcout << "Convergence rate: " << std::fixed << std::setprecision(2)
+          << conv_rate << std::endl;
+  }
+  else
+  {
+    dealii::SolverControl solver_control(solution.size(), 1.e-6, true, true);
```

Comment:
Make this tolerance a command line argument.

Reply:
Why do we need even more command line arguments? It's really difficult to keep track of all the possible ways to provide parameters already. I would actually prefer if we hardcode the configuration for every test file instead of using `hierarchy_input.info` at all.

Reply:
Because it would be helpful to study number of iterations as a function of tolerance. It would be fine to have it as a parameter in `hierarchy_input.info`.